File: src/dotnet/Ansight.Tools.FileSystem/FileSystemOptionsBuilderExtensions.cs
namespace Ansight.Tools.FileSystem;

using System;

public static class FileSystemOptionsBuilderExtensions
{
    public static Options.OptionsBuilder WithFileSystemTools(this Options.OptionsBuilder builder)
        => builder.WithFileSystemTools(static _ => { });

    public static Options.OptionsBuilder WithFileSystemTools(
        this Options.OptionsBuilder builder,
        Action<FileSystemToolsOptionsBuilder> configure)
    {
        ArgumentNullException.ThrowIfNull(builder);
        ArgumentNullException.ThrowIfNull(configure);

        var optionsBuilder = FileSystemToolsOptions.CreateBuilder();
        configure(optionsBuilder);
        var options = optionsBuilder.Build();

        return builder.AddTools(new ITool[]
        {
            new ListDirectoryTool(options),
            new ReadFileTool(options),
            new GetFileChecksumTool(options),
            new DownloadFileTool(options),
            new BeginBinaryDownloadTool(options)
        });
    }
}

File: src/android/ansight-tools-filesystem/src/main/kotlin/ai/ansight/tools/filesystem/FileSystemOptionsBuilderExtensions.kt
package ai.ansight.tools.filesystem

import ai.ansight.runtime.AnsightOptionsBuilder

fun AnsightOptionsBuilder.withFileSystemTools(): AnsightOptionsBuilder {
    return addTools(AndroidFileSystemTools.create())
}

fun AnsightOptionsBuilder.withFileSystemTools(
    options: AndroidFileSystemToolsOptions,
): AnsightOptionsBuilder {
    return addTools(AndroidFileSystemTools.create(options))
}

fun AnsightOptionsBuilder.withFileSystemTools(
    configure: AndroidFileSystemToolsOptionsBuilder.() -> Unit,
): AnsightOptionsBuilder {
    val builder = AndroidFileSystemToolsOptions.createBuilder()
    builder.configure()
    return withFileSystemTools(builder.build())
}
